### PR TITLE
Add ref/unref to timers for Node.js

### LINKED ISSUE #336: After Zone.js patches things setTimeout(...).unref() doesn't work in Node.js
From https://github.com/angular/zone.js/blob/master/lib/node/node.ts#L16, I've inserted two lines that show the problem:

```js
setTimeout(function () {}, 50).unref() // works
if (shouldPatchGlobalTimers) {
  timers_1.patchTimer(_global, set, clear, 'Timeout');
  timers_1.patchTimer(_global, set, clear, 'Interval');
  timers_1.patchTimer(_global, set, clear, 'Immediate');
}
setTimeout(function () {}, 50).unref() // throws error
```

I'm importing `angular2-universal/polyfills` which imports `zone.js/dist/zone-node.js`.
Using Node v4.4.3.

### PR DESCRIPTION
**Pull request by matthewloring**
Fixes #336

### CODE REVIEW

**Comment by matthewloring** [comment]
cc @mhevery @ofrobots @jmdobry

File: test/common/setInterval.spec.ts [hunk 1]
```diff
@@ -1,6 +1,12 @@
 'use strict';
 import {zoneSymbol} from "../../lib/common/utils";
 
+var isNode = true;
```
Comment (mhevery) [inline test/common/setInterval.spec.ts]: 1) This is a better test `isNode = typeof window !== 'undefined'`.
2) could we add this to a separate file and import it from there?

File: test/common/setInterval.spec.ts [hunk 2]
```diff
@@ -25,6 +31,9 @@ describe('setInterval', function () {
       };
       expect(Zone.current.name).toEqual(('TestZone'));
       cancelId = setInterval(intervalFn, 10);
+      if (isNode) {
+        expect(typeof cancelId.ref).toEqual(('function'));
```
Comment (mhevery) [inline test/common/setInterval.spec.ts]: also add test for `unref`

File: test/common/setTimeout.spec.ts [hunk 3]
```diff
@@ -1,5 +1,11 @@
 import {zoneSymbol} from '../../lib/common/utils';
 
+var isNode = true;
+try {
```
Comment (mhevery) [inline test/common/setTimeout.spec.ts]: same heer.

File: test/common/setTimeout.spec.ts [hunk 4]
```diff
@@ -22,6 +28,9 @@ describe('setTimeout', function () {
       };
       expect(Zone.current.name).toEqual(('TestZone'));
       cancelId = setTimeout(timeoutFn, 3);
+      if (isNode) {
+        expect(typeof cancelId.ref).toEqual(('function'));
```
Comment (mhevery) [inline test/common/setTimeout.spec.ts]: also `unref`

File: lib/common/timers.ts [hunk 5]
```diff
@@ -36,7 +36,17 @@ export function patchTimer(
...
       };
-      return zone.scheduleMacroTask(setName, args[0], options, scheduleTask, clearTask);
+      var task = zone.scheduleMacroTask(setName, args[0], options, scheduleTask, clearTask) ;
+      if (!task) {
+        return task;
+      }
+      // Node.js must additionally support the ref and unref functions.
+      var handle = (<TimerOptions>task.data).handleId;
+      if ((<any>handle).ref && (<any>handle).unref) {
+        (<any>task).ref = (<any>handle).ref.bind(handle);
```
Comment (mhevery) [inline lib/common/timers.ts]: What do these functions do? Do they need to be wrapped with zone specific logic?

File: test/common/setInterval.spec.ts (hunk 1 above)
Comment (vicb) [inline test/common/setInterval.spec.ts]: wrt to 1 it will not work in web workers, right ?

File: test/common/setTimeout.spec.ts (hunk 4 above)
Comment (matthewloring) [inline test/common/setTimeout.spec.ts]: Done.

File: test/common/setInterval.spec.ts (hunk 2 above)
Comment (matthewloring) [inline test/common/setInterval.spec.ts]: Done.

File: test/common/setInterval.spec.ts (hunk 1 above)
Comment (matthewloring) [inline test/common/setInterval.spec.ts]: 1) I am not familiar with the web platform or web workers. I can replace this with any check that is desired.
2) Where is the best place for this to live? Should it get its own file?

File: lib/common/timers.ts (hunk 5 above)
Comment (matthewloring) [inline lib/common/timers.ts]: They determine whether the existence of a pending timer should keep the process alive so work can be completed. This should be independent of zones behavior and should not require zone logic.

File: test/common/setInterval.spec.ts (hunk 1 above)
Comment (mhevery) [inline test/common/setInterval.spec.ts]: I think this should do the trick:

1) `var isNode = typeof global == 'object';`
2) We should import `isNode` from `../../lib/common/utils`

File: test/common/setTimeout.spec.ts (hunk 3 above)
Comment (matthewloring) [inline test/common/setTimeout.spec.ts]: Done.

File: test/common/setInterval.spec.ts (hunk 1 above)
Comment (matthewloring) [inline test/common/setInterval.spec.ts]: Done.

**Comment by takeshi** [comment]
I want to apply this patch to use @google/cloud-debug with zone.js on GAE  Node.JS. 
But this patch is not published to npm repository.  Please publish it.

**Pull request merged**